packages/web3-providers/src/Approval/index.ts

```diff
@@ -0,0 +1,176 @@
+
+            mapKeys(nftSpenderRecords, (spender, spenderAddress) =>
+                mapKeys(spender, (token, tokenAddress) => {
+                    const latestTx = maxBy(token, (x) => x.blockNumber)
+                    if (!latestTx) return
+                    const approved = Number.parseInt(latestTx.data, 16) === 1
+                    approvalListState.updateNFT_State(
+                        account,
+                        spenderAddress,
+                        tokenAddress,
+                        chainId,
+                        toBlock,
+                        latestTx.blockNumber,
+                        approved,
+                    )
+                }),
+            )
+
+            const spenderList = approvalListState.nftState[account]?.[chainId]?.spenderList
+
+            if (!spenderList) return EMPTY_LIST
+
+            return Object.keys(spenderList)
+                .map((spender) => {
+                    return Object.keys(spenderList[spender]).map((address) => {
+                        const maskDappContractInfo = maskDappContractInfoList.find((y) =>
+                            isSameAddress(y.address, spender),
+                        )
+
+                        return {
+                            contract: { address, name: '' },
+                            address: spender,
+                            name: maskDappContractInfo?.name,
+                            logo: maskDappContractInfo?.logo,
+                            amount: '1',
+                            approved: spenderList[spender][address].approved,
+                            transactionBlockNumber: spenderList[spender][address].transactionBlockNumber,
+                            isMaskDapp: Boolean(maskDappContractInfo),
+                        }
+                    })
+                })
+                .flat()
+                .filter((x) => x.approved)
+                .sort((a, b) => {
+                    if (a.isMaskDapp && !b.isMaskDapp) return -1
+                    if (!a.isMaskDapp && b.isMaskDapp) return 1
+                    return Number(b.transactionBlockNumber) - Number(a.transactionBlockNumber)
+                }) as Array<NonFungibleContractSpender<ChainId, SchemaType>>
+        } catch (error) {
+            console.log(error)
+            return EMPTY_LIST
+        }
+    }
+}
+
+function parseLogs(web3: Web3, logs: Log[]) {
+    return logs
+        .filter((x) => x.topics.length === 3 && x.data !== '0x')
+        .reduce<Record<string, Record<string, Array<{ blockNumber: number; data: string }>>>>((acc, cur) => {
+            const spender = web3.eth.abi.decodeParameter('address', cur.topics[2]) as unknown as string
+            if (isZeroAddress(spender)) return acc
+            if (!acc[spender]) acc[spender] = {}
+
+            if (acc[spender][cur.address]) {
+                acc[spender][cur.address].push({ blockNumber: cur.blockNumber, data: cur.data })
```

Comment:
dedup?

Reply:
No need.

Reply:
Sort by `transactionIndex` if in the same block.